function qH(){}
function wH(){}
function zH(){}
function JH(){}
function MH(){}
function mI(){}
function pI(){}
function dI(){qV()}
function TH(){qV()}
function Zu(a,b){a.Vc(b)}
function vbc(a,b){return ncc(a.i,b)}
function aI(){Fb(this,cI(new dI))}
function gI(){Fb(this,iI(new jI))}
function DH(){Fb(this,FH(new GH))}
function QH(){Fb(this,SH(new TH))}
function AH(){AH=Qzc;vH=new zH}
function NH(){NH=Qzc;IH=new MH}
function qI(){qI=Qzc;lI=new pI}
function jI(){var a;a=(new mI,qI(),lI);oI(a)}
function GH(){var a;a=(new JH,NH(),IH);LH(a);qV()}
function nH(a){this.d=a;this.f=sH(new tH(this));kH(this)}
function Fcc(a){lcc();Ecc.call(this);Acc(this,a)}
function _G(a,b,c){Mx.call(this,a,b);this.a=c}
function ZH(){Fb(this,OTb($H(OBc+(new sV,FV(),qV(),_Jc)+OBc).a))}
function iI(){var a;a=OTb(rI(OBc+(new sV,FV(),qV(),_Jc)+OBc).a);return a}
function cI(){var a;a=OTb(eI(OBc+(FV(),qV(),_Jc)+OBc,aKc,bKc,WRb((xV(),dV).d.a),cKc,IJc).a);return a}
function FH(){var a;a=OTb(OH(OBc+(FV(),qV(),_Jc)+OBc,aKc,bKc,WRb((xV(),dV).d.a),cKc,IJc).a);return a}
function OTb(a){var b;NTb();Y7(LTb,a);b=e8(LTb);N7(b.parentNode,b);return b}
function LH(a){if(!a.a){a.a=true;Hbb();O3(Ebb,OBc);Lbb();return true}return false}
function oI(a){if(!a.a){a.a=true;Hbb();O3(Ebb,OBc);Lbb();return true}return false}
function E6b(a,b){A6b();D6b.call(this,k8($doc,hDc));_3b(this.b,a,false);C6b(this,b)}
function tH(a){var b;this.c=a;b=(new wH,AH(),vH);yH(b);qV();this.a=A8($doc);this.b=new KTb(this.a)}
function ncc(a,b){var c;c=new Fcc(b);(!!c.g||!!c.j)&&(c.g?vcc(c.g,c):!!c.j&&Obc(c.j,c));tcc(a,qcc(a),c);return c}
function XH(){if(!WH){WH=new fxc;etc(WH,OJc,new ZH);etc(WH,yCc,new aI);etc(WH,hCc,new DH);etc(WH,SJc,new QH);etc(WH,'tutorial',new gI)}}
function lH(a,b,c){var d;if(!c){if(a.e.b){if(Ec(a.e.b.k,b))return}c=Vbc(a.e)}while(c.b<c.d.Pe()){d=rnb(ouc(c),194);if(Ec(d.k,b)){Qbc(a.e,d,false);return}}}
function BH(a,b){var c;c=new Sqc;Q5(c.a,LCc);Mqc(c,KRb(a));Q5(c.a,"'><span>Navigation<\/span><\/div> <span id='");Mqc(c,KRb(b));Q5(c.a,$Jc);return new qRb(V5(c.a))}
function yH(a){if(!a.a){a.a=true;Hbb();O3(Ebb,'.gwt-TreeItem-selected a{text-decoration:none;color:black;font-weight:bold;}.gwt-Tree{margin-top:10px;}');Lbb();return true}return false}
function mH(a,b,c,d){var e,f,g;g=c;d!=OBc&&(g=c+YGc+d);lH(a,g,null);Abc(a.e);if(Xpc(c,a.b)){a.c=d}else{e=new a6b(OBc);Kb(e,(FV(),qV(),'GAGBG34BEE'));K7(e.ab,b.ab);JX(a.a,e);a.b=c;a.c=d}if(Xpc(a.c,OBc)){IX(a.a).ab.scrollTop=0}else{f=oZb(a.b+YGc+a.c);!!f&&i8(f)}}
function wv(a){var b;!a.I&&(a.I=(b=new _G((!a.gb&&(a.gb=new Meb),a.gb),(!a.J&&(a.J=new nH((new qH,!a.pb&&(a.pb=ov(a)),a.pb),!a.fb&&(a.fb=new sV))),a.J),(!a.H&&(a.H=new dH),!a.pb&&(a.pb=ov(a)),a.pb),!a.K&&(a.K=new XH)),Zu(b,(!a.lb&&(a.lb=new vkc),a.lb)),b));return a.I}
function sH(a){var b,c,d,e,f;b=new O4b((_ab(),Tab));K4b(b,(c=new a6b(BH(OBc+(FV(),qV(),XJc)+OBc,a.a).a),Bb(c,YJc),d=MTb(c.ab),JTb(a.b),d.b?M7(d.b,d.a,d.c):PTb(d.a),$5b(c,(f=new Xbc,a.c.e=f,f),JTb(a.b)),c),17);N4b(b,(e=new LX,Bb(e,ZJc),a.c.a=e,e),(a5b(),V4b),0);return b}
function SH(){var a;a=OTb(UH(OBc+(FV(),qV(),_Jc)+OBc,aKc,bKc,WRb((xV(),dV).d.a),cKc,IJc,aKc,bKc,WRb(dV.d.a),cKc,IJc,aKc,bKc,WRb(dV.d.a),cKc,IJc,aKc,bKc,WRb(dV.d.a),cKc,IJc,aKc,bKc,WRb(dV.d.a),cKc,IJc,aKc,bKc,WRb(dV.d.a),cKc,IJc,aKc,bKc,WRb(dV.d.a),cKc,IJc,aKc,bKc,WRb(dV.d.a),cKc,IJc,aKc,bKc,WRb(dV.d.a),cKc,IJc,aKc,bKc,WRb(dV.d.a),cKc,IJc,aKc,bKc,WRb(dV.d.a),cKc,IJc,aKc,bKc,WRb(dV.d.a),cKc,IJc,aKc,bKc,WRb(dV.d.a),cKc,IJc).a);return a}
function kH(a){var b,c,d,e,f;new Gcc(ZCc);f=new smc(ECc);vbc(a.e,new E6b(QJc,$l(a.d,new rmc(f,NJc,OJc)))).k=OJc;e=new Fcc(new E6b(RJc,$l(a.d,new rmc(f,NJc,yCc))));e.k=yCc;ncc(e,new E6b('Uploading Phenotypes',$l(a.d,new rmc(new rmc(f,NJc,yCc),PJc,'upload')))).k='phenotypes_upload';ncc(e,new E6b('Phenotype Format',$l(a.d,new rmc(new rmc(f,NJc,yCc),PJc,'format')))).k='phenotypes_format';ubc(a.e,e);b=new Fcc(new E6b(WCc,$l(a.d,new rmc(f,NJc,hCc))));b.k=hCc;ubc(a.e,b);c=new Fcc(new E6b(YCc,$l(a.d,new rmc(f,NJc,SJc))));c.k=SJc;ncc(c,new E6b('Navigation',$l(a.d,new rmc(new rmc(f,NJc,SJc),PJc,TJc)))).k='analysis_navigation';ncc(c,new E6b(RJc,$l(a.d,new rmc(new rmc(f,NJc,SJc),PJc,yCc)))).k='analysis_phenotypes';ncc(c,new E6b(UJc,$l(a.d,new rmc(new rmc(f,NJc,SJc),PJc,zCc)))).k='analysis_datasets';ncc(c,new E6b(VJc,$l(a.d,new rmc(new rmc(f,NJc,SJc),PJc,wCc)))).k='analysis_transformations';ncc(c,new E6b(WJc,$l(a.d,new rmc(new rmc(f,NJc,SJc),PJc,'results')))).k='analysis_results';ubc(a.e,c);d=Vbc(a.e);while(d.b<d.d.Pe()){ycc(rnb(ouc(d),194),true,false)}}
function $H(a){var b;b=new Sqc;Q5(b.a,LCc);Mqc(b,KRb(a));Q5(b.a,"'><h1>Introduction<\/h1> <p> GWAPP is an interactive web-based application for conducting <a href='http://en.wikipedia.org/wiki/Genome-wide_association_study' target='_blank'>Genome Wide Assocation Studies (GWAS)<\/a> in <a href='http://en.wikipedia.org/wiki/Arabidopsis_thaliana' target='_blank'>Arabidopsis thaliana<\/a>. <br> The only requirement for using GWAPP is a modern browser (supports HTML5). Users can upload phenotypes measured for a subset of 1386 publicly available accessions (see <a href='http://gwas.gmi.oeaw.ac.at/index.html#accessionPage'>accession list<\/a> for details). The uploaded phenotypes are private to the user and are uniquely identified by a dataset key.<br> GWAPP features an extensive, interactive, and a user-friendly interface that includes interactive manhattan plots and interactive local and genome-wide LD plots.  It facilitates exploratory data analysis by implementing features such as the inclusion of candidate SNPs in the model as cofactors. <\/p><\/div>");return new qRb(V5(b.a))}
function rI(a){var b;b=new Sqc;Q5(b.a,LCc);Mqc(b,KRb(a));Q5(b.a,"'><h1>Tutorial<\/h1> <p>This tutorial shows the typical steps in doing a Genome Wide Association Mapping analysis. Following steps are going to be described in detail: <ol> <li>Upload phenotypes<\/li> <li>Create datasets<\/li> <li>Create transformations<\/li> <li>Run GWAS<\/li> <li>View the results<\/li> <li>Run Step-Wise GWAS<\/li> <\/ol> <\/p> <h2>Uploading Phenotypes<\/h2> <p> First the user has to navigate to the phenotype-upload page via the main navigation bar. In this tutorial we are going to upload a phenotype file which contains 4 different phenotypes. For a detailed description of the phenotype file format refer to <a href='#!helpPage;section=phenotypes;subsection=format'>Phenotype Format section<\/a>. Just click on \"Browse\" in the upload-form and select your phenotype. <img> As soon as the phenotypes were successfully uploaded the app will switch to the Analysis page which displays the list of all phenotypes and some statistics. <img> <\/p> <h2> Creating a subset <\/h2> <p> By clicking on a specific phenotype additional information about the selected phenotype are displayed. On this page also all subsets/datasets for the phenotype are displayed. Initially after uploading a phenotype a dataset called fullset whichc contains phenotype values for all accessions is created. It is now possible to create subsets for specific accessions. For example if the users wants to run GWAS on a swedish subset it is possible to <\/p><\/div>");return new qRb(V5(b.a))}
function OH(a,b,c,d,e,f){var g;g=new Sqc;Q5(g.a,LCc);Mqc(g,KRb(a));Q5(g.a,"'><h1>Accessions<\/h1> <p> This page shows information about all the 1307 Arabidopsis Thaliana accessions for which the GWAS-Web-App contains genotype information and can be used for GWAS studies. <\/p> <p> The page consists of 4 sections (see figure): <\/p> <div class='");Mqc(g,KRb(b));Q5(g.a,"' id='accessions'> <div> <a class='");Mqc(g,KRb(c));Q5(g.a,"' href='#!helpPage;section=accessions'><img src='");Mqc(g,KRb(d.a));Q5(g.a,"'><\/a> <a href='#accessions'> <img class='");Mqc(g,KRb(e));Q5(g.a,"' src='/public/resources/screens/accessions.png'> <\/a> <i>Figure: Accession-View (click for fullscreen)<\/i> <\/div> <\/div> <div class='");Mqc(g,KRb(f));Q5(g.a,"'><\/div> <ul> <li><b>Filter-Box<\/b>: Filtering the accession list<\/li> <li><b>Google Map<\/b>: Location information of each accession<\/li> <li><b>Geographic Distribution Map<\/b>: Overview of the geographic distribution of the accessions<\/li> <li><b>Accession-List<\/b>: List of the accessions<\/li> <\/ul> <p> The Filter-Box allows to filter the <b>Accession-List<\/b> according to <b>accession-name<\/b>, <b>country<\/b> and/or <b>collector<\/b>. <br> By clicking on any accession in the <b>Accession-List<\/b> its location is shown in the <b>Google Map<\/b>. <br> By clicking on any accession in the google map the accession-name and id is shown in a popup. <br> By moving the mouse over the <b>Geographic Distribution Map<\/b> the number of accessions for each country is shown. <\/p><\/div>");return new qRb(V5(g.a))}
function eI(a,b,c,d,e,f){var g;g=new Sqc;Q5(g.a,LCc);Mqc(g,KRb(a));Q5(g.a,"'><h1>Phenotypes<\/h1> <p> In order to use the GWAS-Web-App the user first has to upload his/her phenotypes. The phenotypes can be uploaded by navigating to the <a href='#!phenotypeUploadPage'>Upload Phenotypes<\/a> page.<br> The user has to make sure that the phenotypes are avaialble in the correct format (<a href='/public/resources/phenotypes.csv'>Download<\/a> sample phenotype-file).  <\/p> <h2><a id='phenotypes_upload' name='uploading_phenotypes'>Uploading Phenotypes<\/a><\/h2> <p>Uploading phenotypes is quite straightforward.<br> The user navigates to the <a href='#!phenotypeUploadPage'>Upload Phenotypes<\/a> page and can either browse for a phenotype file or paste the contents of a phenotype file into the <b>Content textarea<\/b> By pressing the upload button the phenotypes will be uploaded to the server and stored under the users credentials.<br> <\/p> <div class='");Mqc(g,KRb(b));Q5(g.a,"' id='upload_phenotype'> <div> <a class='");Mqc(g,KRb(c));Q5(g.a,"' href='#!helpPage;section=phenotypes'><img src='");Mqc(g,KRb(d.a));Q5(g.a,"'><\/a> <a href='#upload_phenotype'> <img class='");Mqc(g,KRb(e));Q5(g.a,"' src='/public/resources/screens/upload_phenotypes.png'> <\/a> <i>Figure: Upload Phenotypes page (click figure for fullscreen)<\/i> <\/div> <\/div> <div class='");Mqc(g,KRb(f));Q5(g.a,"'><\/div> <p> If the uploading was successful the user will be automatically redirected to the <a href='#!helpPage;subsection=phenotypes;section=analysis'>Phenotype-View<\/a> <\/p> <h2><a id='phenotypes_format' name='phenotype_format'>Phenotype format<\/a><\/h2> <p> In order to successfully upload the phenotypes, the phenotypes have to be in the right format.<br> The phenotype file must have at least two columns. The first column contains the accession-ids/ecotype-ids and the second column contains the phenotype-values.<br> Multiple phenotypes can be uploaded by adding additional columns at the end of the first phenotype column.<br> The general format looks like this: <table border='0' cellpadding='2px'> <thead> <tr> <th>ecotype_id<\/th><th>phenotype1<\/th><th>phenotype2<\/th><th>phenotypeN<\/th> <\/tr> <\/thead> <tbody> <tr> <td>integer<\/td><td>float or NA<\/td><td>float or NA<\/td><td>float or NA<\/td><\/tr> <\/tbody> <\/table> There are some important criterias which have to be fulfilled: <ul> <li>There must be a header-line which contains the phenotype names<\/li> <li>Each phenotype-name must be unique<\/li> <li>All ecotype-ids must be part of the 1307 accession list<\/li> <li>The column sperator must be a column (\"<b>,<\/b>\")<\/li> <li>Every row must be returned by a normal carriage return (\"<b>\\n<\/b>\")<\/li> <li>Missing phenotype values must be filled with the string \"<b>NA<\/b>\"<\/li> <\/ul> The following example shows the beginning of a phenotype file which contains 4 phenotypes (LD, FT10, FT16 and FRI): <pre>\n\t\tecotype_id,LD,FT10,FT16,FRI\n\t\t1,NA,65.8,72.0,NA\n\t\t2,NA,74.0,101.25,NA\n\t\t15,NA,59.0,47.6667,NA\n\t\t23,NA,49.0,67.0,NA\n\t\t66,NA,44.0,36.5,NA\n\t\t78,NA,45.0,33.5,NA\n\t\t<\/pre> <a href='/public/resources/phenotypes.csv'>Download<\/a> sample phenotype-file with all 1307 accession-ids and one empty phenotype column. <\/p><\/div>");return new qRb(V5(g.a))}
function UH(a,b,c,d,e,f,g,i,j,k,n,o,p,q,r,s,t,u,v,w,x,y,z,A,B,C,D,E,F,G,H,I,J,K,L,M,N,O,P,Q,R,S,T,U,V,W,X,Y,Z,$,ab,bb,cb,db,eb,fb,gb,hb,ib,jb,kb,lb,mb,nb,ob,pb){var qb;qb=new Sqc;Q5(qb.a,LCc);Mqc(qb,KRb(a));Q5(qb.a,"'><h1><a name='analysis'>Analysis<\/a><\/h1> <p> The Analysis page is the main page in the GWAS-Web-App.<br> It contains most of the functionality and will be used most of the time. It consists of 4 distinct views: <\/p> <ul> <li><b>Phenotype-View<\/b>: information about the phenotypes<\/li> <li><b>Dataset-View<\/b>: information about the datasets<\/li> <li><b>Transformation-View<\/b>: information about the transformations<\/li> <li><b>Results-View<\/b>: GWAS results and plots<\/li> <\/ul> <p> The views are hierarchically structured starting with the Phenotype-View being the root view. Each Phenotype can contain multiple datasets.<br> By default every uploaded phenotype has a <b>Fullset<\/b> dataset which contains all accession for which phenotype-values exists. <br> Each dataset can have multiple transformations. By default every dataset has the <b>raw<\/b> transformation which are the untransformed phenotype-values. Each transformation can have multiple GWAS-results. There are currently 3 different GWAS-methods (EMMAX,LM and KW).  <\/p> <h2><a id='analysis_navigation' name='analysis_navigation'>Navigation<\/a><\/h2> <div class='");Mqc(qb,KRb(b));Q5(qb.a,"' id='navigation'> <div> <a class='");Mqc(qb,KRb(c));Q5(qb.a,dKc);Mqc(qb,KRb(d.a));Q5(qb.a,"'><\/a> <a href='#navigation'> <img class='");Mqc(qb,KRb(e));Q5(qb.a,"' src='/public/resources/screens/navigation.png'> <\/a> <i>Figure: Navigation (click figure for fullscreen)<\/i> <\/div> <\/div> <div class='");Mqc(qb,KRb(f));Q5(qb.a,"'><\/div> <p> Navigation in the view is quite straightforward.<br> On the left side there is a navigation bar containing a tree which shows provides access to the above mentioned 4 views.<br> On the right side the corresponding view will be displayed. <br> <\/p> <h2><a id='analysis_phenotypes' name='analysis_phenotypes'>Phenotypes<\/a><\/h2> <p> The Phenotype-View consists of two different sub-views. By default (when clicking on Analysis) the phenotype-list view is loaded (see <a href='#navigation'>Navigation-Figure<\/a>).<Br><\/Br> The phenotype-list view shows general information about all currently stored phenotypes and also contains two barcharts. One barchart visualizes the number of phenotype values and the other visualizes the number of datasets, transformations and GWAS-results contained in a phenotype.<br> The user can also delete a specific phenotype by clicking on the delete button.<br> As soon as the user selects a specific phenotype from the navigation tree on the left side, the Phenotype-View is loaded (see figure below) <\/p> <div class='");Mqc(qb,KRb(g));Q5(qb.a,"' id='phenotype_view'> <div> <a class='");Mqc(qb,KRb(i));Q5(qb.a,dKc);Mqc(qb,KRb(j.a));Q5(qb.a,"'><\/a> <a href='#phenotype_view'> <img class='");Mqc(qb,KRb(k));Q5(qb.a,"' src='/public/resources/screens/phenotype_view.png'> <\/a> <i>Figure: Phenotype-View (click figure for fullscreen)<\/i> <\/div> <\/div> <div class='");Mqc(qb,KRb(n));Q5(qb.a,"'><\/div> <p> In the top area of the phenotype-view general phenotype information are displayed.<br> Below a list of all datasets in the specific phenotype are shown.<br> The <b>Geographic Distribution Map<\/b> for the selected dataset in that list is displayed further down.<br> By selecting another dataset from the dataset table this <b>Geographic Distribution Map<\/b> will get updated. <\/p> <p> The dataset table also allows to delete a dataset (with the exception of the default <b>Fullset<\/b>)<br> By clicking on the <b>Subset<\/b> button it is possible to create a new dataset based on the selected dataset (more about that in the next section). In the bottom part of the view some statistics about the datasets contained in the phenotype are displayed. On the left side the number of accessions/values in each dataset is displayed and on the right side the number of transformations and different GWAS-results are visualized.  <\/p> <h2><a id='analysis_datasets' name='analysis_datasets'>Datasets<\/a><\/h2> <p> By selecting a specific dataset from the left navigation tree or clicking on the <b>Subset<\/b> button the Dataset-View is loaded.<br> The Dataset-View consists of different information-boxes: <\/p> <p> <ul> <li><b>Filter-Box<\/b>: This box allows to filter the accession-list. In Edit-Mode it is possible to change the dataset name and description.<\/li> <li><b>Accession-List<\/b>: This table shows all accessions which are currently forming the dataset. In Edit-Mode the accession-list is used to add and remove accessions from the dataset.<\/li> <li><b>Geographic Distribution Map<\/b>: Visualizes the geographic distribution of the dataset<\/li> <li><b>Google Map<\/b>: Shows the locations of all accessions in the dataset on a world map<\/li> <\/ul> <\/p> <div class='");Mqc(qb,KRb(o));Q5(qb.a,"' id='dataset_view'> <div> <a class='");Mqc(qb,KRb(p));Q5(qb.a,dKc);Mqc(qb,KRb(q.a));Q5(qb.a,"'><\/a> <a href='#dataset_view'> <img class='");Mqc(qb,KRb(r));Q5(qb.a,"' src='/public/resources/screens/create_subsets.png'> <\/a> <i>Figure: Dataset-View (click figure for fullscreen)<\/i> <\/div> <\/div> <div class='");Mqc(qb,KRb(s));Q5(qb.a,"'><\/div> <p> In <b>Readonly mode<\/b> the user can filter/search the accession list.<br> In <b>Edit mode<\/b> (either by editing a dataset or creating a new one) the user can remove or add accessions to the dataset.<br> Only accessions from the fullset can be added or removed from the dataset. The <b>Fullset CAN NOT<\/b> be deleted or edited. <\/p> <p> Removing or adding accessions to a dataset is down by clicking on the checkbox in the accession-list. Adding or removing multiple accessions can be done by pressing the <b>SHIFT<\/b> keyboard-button while clicking on the checkboxes. <\/p> <h2><a id='analysis_transformations' name='transformations'>Transformations<\/a><\/h2> <p> Selecting a transformation from one of the datasets will load the Transformation-View.<br> The transformation view contains following information widgets: <ul> <li><b>Transformation-list<\/b>: list with all transformations in the dataset<\/li> <li><b>Histogram<\/b>: histogram of the transformed phenotype-values<\/li> <li><b>Phenotype-Explorer<\/b>: motionchart of phenotype-values and accession information<\/li> <li><b>Results-list<\/b>: list of all GWAS-results for the currently selected transformation<\/li> <\/ul> <\/p> <div class='");Mqc(qb,KRb(t));Q5(qb.a,"' id='transformation_view'> <div> <a class='");Mqc(qb,KRb(u));Q5(qb.a,dKc);Mqc(qb,KRb(v.a));Q5(qb.a,"'><\/a> <a href='#transformation_view'> <img class='");Mqc(qb,KRb(w));Q5(qb.a,"' src='/public/resources/screens/transformation_view.png'> <\/a> <i>Figure: Transformation-View (click figure for fullscreen)<\/i> <\/div> <\/div> <div class='");Mqc(qb,KRb(x));Q5(qb.a,"'><\/div> <p> The transformation-list shows all transformations in the current selected dataset.<br> Users can delete transformations, create new ones and also run one of the GWAS-method on a specific transformation.<\/p> <p> Below the histogram of the phenotype-values for the selected transformations is displayed.<br> In case the user creates a new transformation, a preview-histogram for the desired transformation is shown in addition to the original histogram.<br> <\/p> <p> Further down the Phenotype-Explorer is shown. <\/p> <h3>Phenotype-Explorer<\/h3> <p> The phenotype explorer has two modes. First it can display a scatterplot/xy-plot of all accessions. The x- and y-axis can be set to longitude, latitude or phenotype. Additional information ca be visualized by the size and the color of the circles (see <a href='#phenotype_explorer_scatter'>figure<\/a> below). <\/p> <div class='");Mqc(qb,KRb(y));Q5(qb.a,"' id='phenotype_explorer_scatter'> <div> <a class='");Mqc(qb,KRb(z));Q5(qb.a,dKc);Mqc(qb,KRb(A.a));Q5(qb.a,"'><\/a> <a href='#phenotype_explorer_scatter'> <img class='");Mqc(qb,KRb(B));Q5(qb.a,"' src='/public/resources/screens/phenotype_explorer_scatter.png'> <\/a> <i>Figure: Phenotype-Explorer Scatterchart (click figure for fullscreen)<\/i> <\/div> <\/div> <div class='");Mqc(qb,KRb(C));Q5(qb.a,"'><\/div> <p> The second mode is a barchart where the x-axis can be ordered by phenotype, longitude, latitude or alphabetically and the y-axis can bei the actual values. Again color can code another information (see <a href='#phenotype_explorer_bar'>figure<\/a> below). <\/p> <div class='");Mqc(qb,KRb(D));Q5(qb.a,"' id='phenotype_explorer_bar'> <div> <a class='");Mqc(qb,KRb(E));Q5(qb.a,dKc);Mqc(qb,KRb(F.a));Q5(qb.a,"'><\/a> <a href='#phenotype_explorer_bar'> <img class='");Mqc(qb,KRb(G));Q5(qb.a,"' src='/public/resources/screens/phenotype_explorer_bar.png'> <\/a> <i>Figure: Phenotype-Explorer Barchart (click figure for fullscreen)<\/i> <\/div> <\/div> <div class='");Mqc(qb,KRb(H));Q5(qb.a,"'><\/div> <h2><a id='analysis_results' name='results'>Results<\/a><\/h2> <p> The Results-View displays the GWAS-result.<br> It consists of a tab-panel with 2 main views:<\/p> <h3>GWAS-Plots<\/h3> <p> For each chromosome a scatterplot will be displayed showing the <b>top 10.0000 SNPs<\/b>.<br> The x-axis measures the positions along the genome/chromosome and the y-axis displays the p-Value scale.<br> The <b><a href='http://en.wikipedia.org/wiki/False_discovery_rate#Benjamini.E2.80.93Hochberg.E2.80.93Yekutieli_procedure' target='_blank'>Benjamini\u2013Hochberg\u2013Yekutieli<\/a><\/b> multiple testing procedure (Benjamini and Yekutieli, 2001) was used to control the false discovery rate. Assuming arbitrary dependence between SNPs, the 5% FDR threshold is plotted as a dashed horizontal line.<br> In the top left corner there is a autofilled textbox which allows to search for a specific Arabidopsis Thaliana gene.<br> If the user selects a gene it will be highlighted as a colored band in the corresponding scatterplot.<br> By moving the mouse over a SNP the pValue and the position of the SNP will be displayed in the upper right corner of the scatterplot.<br> Clicking on a SNP in the scatterplot opens a popup which allows the user to run a step-wise GWAS analysis with the selected SNP as a co-factor (<b>works only with EMMAX<\/b>)  <\/p> <div class='");Mqc(qb,KRb(I));Q5(qb.a,"' id='results_view_plots'> <div> <a class='");Mqc(qb,KRb(J));Q5(qb.a,dKc);Mqc(qb,KRb(K.a));Q5(qb.a,"'><\/a> <a href='#results_view_plots'> <img class='");Mqc(qb,KRb(L));Q5(qb.a,"' src='/public/resources/screens/results_view_plots.png'> <\/a> <i>Figure: GWAS-plots (click figure for fullscreen)<\/i> <\/div> <\/div> <div class='");Mqc(qb,KRb(M));Q5(qb.a,"'><\/div> <p> Below each scatterplot an invisible <b>Geneviewer<\/b> is hidden which is displayed as soon as the user zooms into a region smaller than <b>1.5 Mb<\/b>.<br> The user can zoom either vertically along the y-xis or horizontally along the x-axis.<br> Zooming in the scatterplot is done by pressing and holding the left mouse button somewhere in the scatterplot (this defines the staring point) and and then dragging the mouse (while holding the left mouse button) to the endpoint of the zoomarea.<br> A grey band visualizes the region which the scatterplot will be zoomed into as soon as the user releases the left mouse-button. Zooming out is done by double clicking somewhere into the scatterplot.<\/p> <p> The <b>Geneviewer<\/b> has two zoom levels.<br> The first zooom level is between <b>1.5 Mb<\/b> and <b>100 kb<\/b> and the Geneviewer will display a simple gene model without any features (see <a href='#results_view_plots_zoom1'>figure<\/a> below).<br> When the user moves the mouse over a gene a popup with information about the specific gene will be displayed.<br> Furthermore moving the mouse over a SNP in the scatterplot will show a vertical line visualizing the position of the SNP in the Geneviewer and highlighting the gene in case the SNP lies on the gene. <\/p> <div class='");Mqc(qb,KRb(N));Q5(qb.a,"' id='results_view_plots_zoom1'> <div> <a class='");Mqc(qb,KRb(O));Q5(qb.a,dKc);Mqc(qb,KRb(P.a));Q5(qb.a,"'><\/a> <a href='#results_view_plots_zoom1'> <img class='");Mqc(qb,KRb(Q));Q5(qb.a,"' src='/public/resources/screens/results_view_plots_zoom1.png'> <\/a> <i>Figure: GWAS-plots and Geneviewer with simple gene model (click figure for fullscreen)<\/i> <\/div> <\/div> <div class='");Mqc(qb,KRb(R));Q5(qb.a,"'><\/div> <p> The second zoom level is displayed when the user zoomed into a region smaller than <b>1.5 Mb<\/b><br> A more complex gene model will be displayed with gene-features (UTRs and exons).<br> Moving the mouse over a SNP in the scatterplot will not only display which gene but also which gene feature was hit. <\/p> <div class='");Mqc(qb,KRb(S));Q5(qb.a,"' id='results_view_plots_zoom2'> <div> <a class='");Mqc(qb,KRb(T));Q5(qb.a,dKc);Mqc(qb,KRb(U.a));Q5(qb.a,"'><\/a> <a href='#results_view_plots_zoom2'> <img class='");Mqc(qb,KRb(V));Q5(qb.a,"' src='/public/resources/screens/results_view_plots_zoom2.png'> <\/a> <i>Figure: GWAS-plots and Geneviewer with complex gene model (click figure for fullscreen)<\/i> <\/div> <\/div> <div class='");Mqc(qb,KRb(W));Q5(qb.a,"'><\/div> <p> In case GWAS result the scatterplots will also indicated which SNPs where used as co-factors in the model by drawing a vertical line to the corresponding SNP. <\/p> <p> Below the <b>GeneViewer<\/b> genome wide statistics are displayed as a histogram (see figure below). By default the binned gene count is displayed. By clicking on the gear symbol a popup opens and users can display various genome wide statistics (i.e. Fst, CLR, etc).  The statistics panel is zoomable in both axis and a yellow band shows the region in the chromosome that corresponds to the zoom level of the Manhatten chart/GeneViewer. Users can also upload their own statistics by providing a simple CSV (format. chr, pos, float).  <div class='");Mqc(qb,KRb(X));Q5(qb.a,"' id='results_genomewide_stats'> <div> <a class='");Mqc(qb,KRb(Y));Q5(qb.a,dKc);Mqc(qb,KRb(Z.a));Q5(qb.a,"'><\/a> <a href='#results_genomewide_stats'> <img class='");Mqc(qb,KRb($));Q5(qb.a,"' src='/public/resources/screens/results_genomewide_stats.png'> <\/a> <i>Figure: Genomewide statistics in the GWAS plot (click figure for fullscreen)<\/i> <\/div> <\/div> <div class='");Mqc(qb,KRb(ab));Q5(qb.a,"'><\/div> <\/p> <h2><a name='LD-Viewer'>LD-Viewer<\/a><\/h2> <p> Users can also visualize the linkage disequilibrium (LD) structure. This can be done by clicking on any SNP and choosing from three different methods: <ul> <li>Show LD in this region<\/li> <li>Calculate exact LD in this region<\/li> <li>Highlight SNPs in LD for this SNP<\/li> <\/ul> The first two options ( \u201CShow LD in this region\u201D and \u201CCalculate exact LD in this region\u201D) display  a LD triangle plot below the gene annotation panel and color code the SNPs in the Manhattan plot (see figure below). <\/p> <div class='");Mqc(qb,KRb(bb));Q5(qb.a,"' id='results_view_ld_triangle'> <div> <a class='");Mqc(qb,KRb(cb));Q5(qb.a,dKc);Mqc(qb,KRb(db.a));Q5(qb.a,"'><\/a> <a href='#results_view_ld_triangle'> <img class='");Mqc(qb,KRb(eb));Q5(qb.a,"' src='/public/resources/screens/results_ld_triangle.png'> <\/a> <i>Figure: LD-triangle plot dispalying pariwise r\xB2 values for 500 SNPs (click figure for fullscreen)<\/i> <\/div> <\/div> <div class='");Mqc(qb,KRb(fb));Q5(qb.a,"'><\/div> <p> The difference between the first two options are that the former only displays r\xB2 values for the visible SNPs and the latter will calculate and show the r\xB2 values of all SNPs regardless if they are displayed or not. Both options display pair-wise r\xB2 values of at most 500 SNPs (due to limitations regarding visualization and computational complexity). For the sake of visual clarity, only r\xB2 value above a certain threshold (0.3) are color coded. Furthermore, selecting a SNP in the Manhattan plot will color code all neighboring SNPs according to their r\xB2 value. At the same time all pair-wise r\xB2 values in the triangle plot will be highlighted. Similarly, when a specific r\xB2 value in the triangle plot is selected the corresponding pair of SNPs in the Manhattan plot are highlighted with the corresponding color coding and the r\xB2 value is displayed on the scale in the triangle plot. Lastly, the third option (\u201CHighlight SNPs in LD for this SNP\u201D) will calculate genome-wide r\xB2 values between the selected SNP and all other displayed SNPs and color code them in the Manhattan plot (see figure below). <\/p> <div class='");Mqc(qb,KRb(gb));Q5(qb.a,"' id='results_view_ld_genomewide'> <div> <a class='");Mqc(qb,KRb(hb));Q5(qb.a,dKc);Mqc(qb,KRb(ib.a));Q5(qb.a,"'><\/a> <a href='#results_view_ld_genomewide'> <img class='");Mqc(qb,KRb(jb));Q5(qb.a,"' src='/public/resources/screens/results_ld_genomewide.png'> <\/a> <i>Figure: Genome Wide LD between selected SNP and all other SNPs (click figure for fullscreen)<\/i> <\/div> <\/div> <div class='");Mqc(qb,KRb(kb));Q5(qb.a,"'><\/div> <h2><a name='statistics'>Statistics<\/a><\/h2> <p> The second view in the Results-View's tab-panel shows statisctics about the GWAS-results.<br> This view is important for the step-wise GWAS results because it shows how some statistical properties change in each step of the step-wise GWAS analysis.<br> This can help to decide when to stop including more co-factors in the GWAS analysis (see introduction for more information).<br> <\/p> <div class='");Mqc(qb,KRb(lb));Q5(qb.a,"' id='results_view_statistics'> <div> <a class='");Mqc(qb,KRb(mb));Q5(qb.a,dKc);Mqc(qb,KRb(nb.a));Q5(qb.a,"'><\/a> <a href='#results_view_statistics'> <img class='");Mqc(qb,KRb(ob));Q5(qb.a,"' src='/public/resources/screens/results_view_statistics.png'> <\/a> <i>Figure: Statistics after 3 steps (click figure for fullscreen)<\/i> <\/div> <\/div> <div class='");Mqc(qb,KRb(pb));Q5(qb.a,"'><\/div> <p> The Statistics view consists of a table with all the statistical indicators/properties at any given step.<br> Futhermore there is a listbox which allows the users to select an indicator for visualization as a line-chart. <\/p><\/div>");return new qRb(V5(qb.a))}
var dKc="' href='#!helpPage;section=analysis'><img src='",cKc='GAGBG34BFE',bKc='GAGBG34BGE',aKc='GAGBG34BHE',_Jc='GAGBG34BIE',QJc='Introduction',RJc='Phenotypes',SJc='analysis',eKc='com.gmi.gwaswebapp.client.mvp.help.sections.',OJc='introduction',NJc='section',PJc='subsection';BMb(251,1,lAc);_.Hc=function Rv(){this.b.bc(wv(this.a.a))};BMb(381,276,oAc,_G);_.Yc=function aH(){var a,b,c;a=dm(this.a);b=omc(a,NJc,OJc);c=omc(a,PJc,OBc);mH(rnb(this.C,67),rnb(_sc(WH,b),195),b,c)};_.$c=function bH(){Hx(this,new Vmc((hJ(),fJ),this))};_.a=null;BMb(384,290,{67:1},nH);_.ed=function oH(){return this.f};_.a=null;_.b=null;_.c=null;_.d=null;_.e=null;_.f=null;BMb(385,1,{},qH);BMb(386,1,{},tH);_.a=null;_.b=null;_.c=null;BMb(387,1,{},wH);var vH=null;BMb(388,1,aAc,zH);_.a=false;BMb(391,5,Uzc,DH);BMb(392,1,{},GH);BMb(393,1,{},JH);var IH=null;BMb(394,1,aAc,MH);_.a=false;BMb(397,5,Uzc,QH);BMb(398,1,{},TH);BMb(400,1,{},XH);var WH=null;BMb(401,5,Uzc,ZH);BMb(404,5,Uzc,aI);BMb(405,1,{},dI);BMb(407,5,Uzc,gI);BMb(408,1,{},jI);BMb(409,1,{},mI);var lI=null;BMb(410,1,aAc,pI);_.a=false;BMb(1254,4,Vzc,E6b);BMb(1310,5,{181:1,194:1,195:1},Fcc);var Zsb=coc(HIc,'HelpPresenter',381),jtb=coc(eKc,'HelpSectionFactory',400),ctb=coc(HIc,'HelpView',384),ktb=coc(eKc,QJc,401),mtb=coc(eKc,RJc,404),gtb=coc(eKc,WCc,391),itb=coc(eKc,YCc,397),qtb=coc(eKc,'Tutorial',407),ltb=coc(eKc,'Phenotypes_PhenotypesUiBinderImpl$Widgets',405),dtb=coc(eKc,'Accessions_AccessionsUiBinderImpl$Widgets',392),htb=coc(eKc,'Analysis_AnalysisUiBinderImpl$Widgets',398),ntb=coc(eKc,'Tutorial_TutorialUiBinderImpl$Widgets',408),btb=coc(HIc,'HelpView_BinderImpl',385),$sb=coc(HIc,'HelpView_BinderImpl$Widgets',386),ftb=coc(eKc,'Accessions_AccessionsUiBinderImpl_GenBundle_default_InlineClientBundleGenerator',393),etb=coc(eKc,'Accessions_AccessionsUiBinderImpl_GenBundle_default_InlineClientBundleGenerator$1',394),ptb=coc(eKc,'Tutorial_TutorialUiBinderImpl_GenBundle_default_InlineClientBundleGenerator',409),otb=coc(eKc,'Tutorial_TutorialUiBinderImpl_GenBundle_default_InlineClientBundleGenerator$1',410),atb=coc(HIc,'HelpView_BinderImpl_GenBundle_default_InlineClientBundleGenerator',387),_sb=coc(HIc,'HelpView_BinderImpl_GenBundle_default_InlineClientBundleGenerator$1',388);JBc(w4)(2);